streaming/src/event_server.h

```diff
@@ -0,0 +1,229 @@
+      full_cv_.wait(lock);
+      STREAMING_LOG(WARNING) << "Event server is full_sleep be notified";
+    }
+    if (!is_started_) {
+      return;
+    }
+    buffer_.push(t);
+    if (1 == Size()) {
+      empty_cv_.notify_one();
+    }
+  }
+
+  void PushToUrgent(const T &t) {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Size() >= capacity_ && is_started_) {
+      STREAMING_LOG(WARNING) << " EventQueue is full, its size:" << Size()
+                             << " all_size:" << Size() << " capacity:" << capacity_
+                             << " buffer size:" << buffer_.size()
+                             << " urgent_buffer size:" << urgent_buffer_.size();
+      full_cv_.wait(lock);
+      STREAMING_LOG(WARNING) << "event_server's full_sleep be notifyed";
+    }
+    if (!is_started_) {
+      return;
+    }
+    urgent_buffer_.push(t);
+    if (Size() == 1) {
+      empty_cv_.notify_one();
+    }
+  }
+
+  void Pop() {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    if (Size() >= capacity_) {
+      STREAMING_LOG(WARNING) << "Pop should notify"
+                             << " all_Size:" << Size();
+    }
+    if (urgent_) {
+      urgent_buffer_.pop();
+    } else {
+      buffer_.pop();
+    }
+    full_cv_.notify_all();
+  }
+
+  bool Get(T &evt) {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Empty() && is_started_) {
+      empty_cv_.wait(lock);
+    }
+    if (!is_started_) {
+      return false;
+    }
+    if (!urgent_buffer_.empty()) {
+      urgent_ = true;
+      evt = urgent_buffer_.front();
+    } else {
+      urgent_ = false;
+      evt = buffer_.front();
+    }
+    return true;
+  }
+
+  T PopAndGet() {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Empty() && is_started_) {
+      empty_cv_.wait(lock);
+    }
+    if (!is_started_) {
+      return T();
+    }
+    if (!urgent_buffer_.empty()) {
+      T res = urgent_buffer_.front();
+      urgent_buffer_.pop();
+      if (Full()) {
+        full_cv_.notify_one();
+      }
+      return res;
+    }
+    T res = buffer_.front();
+    buffer_.pop();
+    if (Size() + 1 == capacity_) full_cv_.notify_one();
+    return res;
+  }
+
+  T &Front() {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    if (urgent_buffer_.size()) {
+      return urgent_buffer_.front();
+    }
+    return buffer_.front();
+  }
+
+  inline bool Empty() { return buffer_.empty() && urgent_buffer_.empty(); }
+
+  inline bool Full() { return buffer_.size() + urgent_buffer_.size() == capacity_; }
+
+  inline size_t Size() { return buffer_.size() + urgent_buffer_.size(); }
+
+  inline size_t UrgentBufferSize() { return urgent_buffer_.size(); }
+
+  inline size_t Capacity() { return capacity_; }
+
+ private:
+  std::mutex ring_buffer_mutex_;
+  std::condition_variable empty_cv_;
+  std::condition_variable full_cv_;
+  std::queue<T> buffer_;
+  std::queue<T> urgent_buffer_;
```

Comment:
could you document the two buffers and how they are used? Is urgent buffer used for high prioritiy requests?